src/distancematrix.rs

```diff
@@ -0,0 +1,220 @@
+//! A distance matrix is stored as a list of taxon names, together with a square
+//! or triangular matrix representing all pairwise distances.
+//!
+//! # Example
+//!
+//! ```rust
+//! use bio_types::distancematrix::DistanceMatrix;
+//!
+//! DistanceMatrix::new(vec!["a".to_string(), "b".to_string()],
+//! vec![
+//!  vec![0.0, 1.0],
+//!  vec![1.0, 0.0],
+//! ]);
+//! ```
+
+use std::{
+    cmp::{max, min},
+    ops::{Index, IndexMut},
+};
+
+use thiserror::Error;
+
+/// The type of the matrix. Either square, lower triangular (excluding the
+/// diagonal), or upper triangular (excluding the diagonal).
+/// Square matrices are assumed to be symmetric, but this is not enforced nor checked.
+#[derive(Debug, PartialEq, Eq)]
+pub enum MatrixType {
+    Square,
+    Lower,
+    Upper,
```

Comment:
Should we also support lower/upper triangular formats that include the diagonal of zeros? I think I've seen it somewhere at some point, but it's hard to find any proper reference for the format sadly.